### PR TITLE
pip module: During --check, if latest is asked, check using pip list --outdated

### LINKED ISSUE #62826: pip module always set to changed when using check mode
##### SUMMARY
The pip module will always be 'changed' when using a virtualenv and when running in check mode.

This is caused by [this block](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/language/pip.py#L472) in the pip module:
```python
def setup_virtualenv(module, env, chdir, out, err):
    if module.check_mode:
        module.exit_json(changed=True)
    ...
```
Instead of actually checking if the virtualenv exists and if all the requirements are present.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
Pip module

##### ANSIBLE VERSION

```paste below
ansible 2.8.5
```

##### CONFIGURATION

```paste below
```

##### OS / ENVIRONMENT

All OS'es

##### STEPS TO REPRODUCE

```yaml
- name: Test pip
  hosts: 127.0.0.1
  connection: local

  pre_tasks:
    - name: Create test dir
      file:
        path: /tmp/test-playbook/
        state: directory

    - name: Create requirements.txt
      copy:
        dest: /tmp/test-playbook/requirements.txt
        content: "ansible==2.8.5"

    - name: Install virtualenv
      pip:
        requirements: /tmp/test-playbook/requirements.txt
        virtualenv: /tmp/test-playbook/venv
        virtualenv_python: python3
```

First execute this playbook with `ansible-playbook test.yml` and then run `ansible-playbook --check test.yml`.

##### EXPECTED RESULTS

The "Install virtualenv" task should not be changed.

##### ACTUAL RESULTS

The task shows that it's changed.

```paste below
/t/test-playbook ❯❯❯ ansible-playbook test.yml --check
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [Test pip] *******************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Create test dir] ************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Create requirements.txt] ****************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Install virtualenv] *********************************************************************************************************************************************************************
changed: [127.0.0.1]

PLAY RECAP ************************************************************************************************************************************************************************************
127.0.0.1                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

Note: when running the playbook without check mode, it is not changed.
```
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [Test pip] *******************************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Create test dir] ************************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Create requirements.txt] ****************************************************************************************************************************************************************
ok: [127.0.0.1]

TASK [Install virtualenv] *********************************************************************************************************************************************************************
ok: [127.0.0.1]

PLAY RECAP ************************************************************************************************************************************************************************************
127.0.0.1                  : ok=4    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### PR DESCRIPTION
**Pull request by JulienPalard**
Mabe fixes #62826

##### SUMMARY

In the pip module, I'm using `pip list --outdated` to list outdated Python packages, in order for `--check` to be able to tell if a package with `state=latest` is to be upgraded or not.

Current situation is to always return `changed=True` when a packaged is `state=latest`, even if it's at the latest version.

The diff may be bigger than absolutely needed:

- During my "understand the code" phase, I noticed `_is_present` takes an unused parameter, so I dropped it, as it's really near my diffs anyway.
- Again during the reading phase, I found the `if module.check_mode` a bit long for what it does (30 lines) itself in a 195 lines of code `main` function, I had hard times understanding it. So I took care of decreasing the complexity instead of increasing it during my implementation by moving a part dedicated to clean and enhance currently installed packaged in the function dedicated to it, and by implementing my code in a dedicated function too. Now the main is 188 lines (a tad shorter), and the `if module.check_mode` is 23 lines (with a feature added!).
- I think there was a potential bug in the `pkg_list = [p for p in out.split('\n')`... line, using the `out` variable instead of the `out_pip` variable, the `out` variable could contain unexpected content from previous runs (setup venv, a few lines before).
- I think there was another potential bug, where `_get_packages` was used elsewhere, and only with a very old pip, as `_get_packages` was not adding `pip` and `setuptools` to the list, if the installation was to install pip or setuptools, it may get unnoticed by `changed = out_freeze_before != out_freeze_after`.

The whole thing clearly lacks tests, I only tested it myself and it worked as expected (--check telling a module should be updated, running the playbook upgrades the module, --check again tells it's OK), but I'm out of time for today and have no idea how to test it for now (any hint appreciated).

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

pip

### CODE REVIEW

**Comment by JulienPalard** [comment]
Python 2.6.

File: lib/ansible/modules/pip.py [hunk 1]
```diff
@@ -350,24 +351,56 @@ def _get_cmd_options(module, cmd):
...
+        # So we need to get those via a specialcase
+        for pkg in ('setuptools', 'pip'):
+            formatted_dep = _get_package_info(module, pkg, env)
+            if formatted_dep is not None:
+                out += '%s\n' % formatted_dep
+
+    # Clean output for pip warnings
+    out = "\n".join([line for line in out.split('\n')
+                     if not line.startswith('You are using')
+                     and not line.startswith('You should consider')])
```
Comment (webknjaz) [inline lib/ansible/modules/pip.py]: you shouldn't need to create an intermediate in-memory list here. Also, `str.startswith()` supports checking multiple values:
```suggestion
    out = "\n".join(
        line for line in out.split('\n')
        if not line.startswith(('You are using', 'You should consider'))
    )
```

File: lib/ansible/modules/pip.py [hunk 2]
```diff
@@ -350,24 +351,56 @@ def _get_cmd_options(module, cmd):
     return cmd_options
 
 
-def _get_packages(module, pip, chdir):
-    '''Return results of pip command to get packages.'''
+def _get_packages(module, pip, chdir, env):
+    '''Return the installed packages as a string, one package per line, like:
```
Comment (webknjaz) [inline lib/ansible/modules/pip.py]: Let's follow PEP 257 here
```suggestion
    '''Return the installed packages as a string.
    
    The result is formatted as one package per line, like:
```

**Comment by ArsenArsen** [comment]
Related to this, `state=latest` outside of check mode never returns changed, despite possibly reinstalling a package.

**Comment by jborean93** [comment]
Closing due to inactivity. If this is still something you wish to pursue then please open a new PR. We expect tests to be added for any changes made that fixes issues or updates the behaviour.

**Pull request closed without merging**